File: src/main/java/com/leijendary/spring/iamtemplate/specification/UserListSpecification.java
package com.leijendary.spring.iamtemplate.specification;

import com.leijendary.spring.iamtemplate.model.IamUser;
import lombok.Builder;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.lang.NonNull;

import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Predicate;
import javax.persistence.criteria.Root;
import java.time.OffsetDateTime;
import java.util.ArrayList;

import static com.leijendary.spring.iamtemplate.util.PredicateUtil.lowerLike;
import static javax.persistence.criteria.JoinType.LEFT;
import static org.apache.commons.lang3.StringUtils.isBlank;

@Builder
public class UserListSpecification implements Specification<IamUser> {

    private final String query;
    private final boolean excludeWithAccounts;
    private final boolean excludeDeactivated;

    @Override
    public Predicate toPredicate(@NonNull final Root<IamUser> root, @NonNull final CriteriaQuery<?> criteriaQuery,
                                 @NonNull final CriteriaBuilder criteriaBuilder) {
        final var predicates = new ArrayList<Predicate>();

        if (!isBlank(query)) {
            // First Name lowercase like
            final var firstName = root.<String>get("firstName");
            final var firstNameLike = lowerLike(query, firstName, criteriaBuilder);

            // Middle Name lowercase like
            final var middleName = root.<String>get("middleName");
            final var middleNameLike = lowerLike(query, middleName, criteriaBuilder);

            // Last Name lowercase like
            final var lastName = root.<String>get("lastName");
            final var lastNameLike = lowerLike(query, lastName, criteriaBuilder);

            // Combine all fields above in an OR criteria
            final var orPredicate = criteriaBuilder.or(firstNameLike, middleNameLike, lastNameLike);

            predicates.add(orPredicate);
        }

        // Option to exclude users with accounts. Normally users with accounts are those who has
        // subscriptions or has multiple users
        if (excludeWithAccounts) {
            final var accountJoin = root.join("account", LEFT);
            final var nullAccount = accountJoin.isNull();

            predicates.add(nullAccount);
        }

        // Option to exclude users that are deactivated
        if (excludeDeactivated) {
            final var deactivatedDate = root.<OffsetDateTime>get("deactivatedDate");
            final var nullDeactivatedDate = deactivatedDate.isNull();

            predicates.add(nullDeactivatedDate);
        }

        // Combine all predicates into a AND criteria
        final var andPredicate = criteriaBuilder.and(predicates.toArray(new Predicate[0]));

        return criteriaQuery.where(andPredicate).getRestriction();
    }
}

File: src/main/kotlin/com/leijendary/spring/template/iam/specification/UserListSpecification.kt
package com.leijendary.spring.template.iam.specification

import com.leijendary.spring.template.iam.api.v1.model.UserExclusionQueryRequest
import com.leijendary.spring.template.iam.core.extension.concat
import com.leijendary.spring.template.iam.core.extension.lowerLike
import com.leijendary.spring.template.iam.entity.Account
import com.leijendary.spring.template.iam.entity.User
import jakarta.persistence.criteria.*
import jakarta.persistence.criteria.JoinType.LEFT
import org.springframework.data.jpa.domain.Specification

class UserListSpecification(
    private val query: String? = null,
    private val exclusionQuery: UserExclusionQueryRequest
) : Specification<User> {
    override fun toPredicate(
        root: Root<User>,
        criteriaQuery: CriteriaQuery<*>,
        criteriaBuilder: CriteriaBuilder
    ): Predicate {
        val predicates = mutableListOf<Predicate>()

        if (!query.isNullOrBlank()) {
            // First name and last name
            val names = listOf<Path<String>>(root.get("firstName"), root.get("lastName"))
            val concatenated = criteriaBuilder.concat(" ", *names.toTypedArray())
            val namesLike = criteriaBuilder.lowerLike(concatenated, query)

            // Email
            val email = root.get<String>("email")
            val emailLike = criteriaBuilder.lowerLike(email, query)

            // (First name and last name) or (email)
            val namesOrEmailLike = criteriaBuilder.or(namesLike, emailLike)

            predicates.add(namesOrEmailLike)
        }

        if (exclusionQuery.exclude.withAccounts) {
            val nullAccount = root.join<Account, User>("account", LEFT).isNull

            predicates.add(nullAccount)
        }

        return criteriaBuilder.and(*predicates.toTypedArray())
    }
}
